src/core_codemods/tempfile_mktemp.py

```diff
@@ -1,44 +1,134 @@
+from textwrap import dedent
+from typing import Optional
+
+import libcst as cst
+from libcst import ensure_type, matchers
+
 from codemodder.codemods.libcst_transformer import (
     LibcstResultTransformer,
     LibcstTransformerPipeline,
 )
-from codemodder.codemods.semgrep import SemgrepRuleDetector
-from codemodder.codemods.utils_mixin import NameResolutionMixin
+from codemodder.codemods.utils_mixin import NameAndAncestorResolutionMixin
+from codemodder.utils.utils import clean_simplestring
 from core_codemods.api import CoreCodemod, Metadata, Reference, ReviewGuidance
 
 
-class TempfileMktempTransformer(LibcstResultTransformer, NameResolutionMixin):
+class TempfileMktempTransformer(
+    LibcstResultTransformer, NameAndAncestorResolutionMixin
+):
     change_description = "Replaces `tempfile.mktemp` with `tempfile.mkstemp`."
     _module_name = "tempfile"
 
-    def on_result_found(self, original_node, updated_node):
-        maybe_name = self.get_aliased_prefix_name(original_node, self._module_name)
-        if (maybe_name := maybe_name or self._module_name) == self._module_name:
-            self.add_needed_import(self._module_name)
-        self.remove_unused_import(original_node)
-        return self.update_call_target(updated_node, maybe_name, "mkstemp")
+    def leave_SimpleStatementLine(self, original_node, updated_node):
+        match original_node:
+            case cst.SimpleStatementLine(body=[bsstmt]):
+                block = self.get_parent(original_node)
+                # SimpleStatementLine is always part of a IndentedBlock or Module body
+                block = ensure_type(block, cst.IndentedBlock | cst.Module)
+                if isinstance(block, cst.Module):
+                    return self.check_mktemp(original_node, bsstmt)
+        return updated_node
+
+    def check_mktemp(
+        self, original_node: cst.SimpleStatementLine, bsstmt: cst.BaseSmallStatement
+    ) -> cst.SimpleStatementLine | cst.FlattenSentinel:
+        if maybe_tuple := self._is_assigned_to_mktemp(bsstmt):  # type: ignore
+            assign_name, call = maybe_tuple
+            return self.report_and_change(call, assign_name)
+        if maybe_tuple := self._mktemp_is_sink(bsstmt):
+            wrapper_func_name, call = maybe_tuple
+            return self.report_and_change(call, wrapper_func_name, assignment=False)
+        return original_node
+
+    def report_and_change(
+        self, node: cst.Call, name: cst.Name, assignment=True
+    ) -> cst.FlattenSentinel:
+        self.report_change(node)
+        self.add_needed_import(self._module_name)
+        self.remove_unused_import(node)
+        with_block = (
+            f"{name.value} = tf.name" if assignment else f"{name.value}(tf.name)"
+        )
+        new_stmt = dedent(
+            f"""
+        with tempfile.NamedTemporaryFile({self._make_args(node)}) as tf:
+            {with_block}
+        """
```

Comment:
This means the file will be closed as soon as it is created. It needs to account for any references after the assignment and include them in the block. 

I remember tackling this problem in the file-resource-leak codemod, so you can refer to it for a partial solution. 
An ideal solution would just include as far as the last reference goes, but a simple one is just to move every statement after the assignment to the `with` block. You can get those by just grabbing all the statements in the parent `cst.Module | cst.IndentedBlock` with index greater than the assignment.

Reply:
I'm trying to understand your comment in the context of the current ` _make_args`. If you look at all unit tests, every one uses `delete=False`. I purposefully didn't account for more complex cases in this iteration. Does that address your concern or did I miss something?

Reply:
Sorry for not being clear, let me try with an example.

Consider the following:
```
tmp_file = open(tempfile.mktemp(), "w+")
tmp_file.write("text")
print(tmp_file.name)
```
If I read the code correctly, the codemod will produce:
```
with tempfile.NamedTemporaryFile("w+") as tf:
    tmp_file = tf.name
tmp_file.write("text")
print(tmp_file.name)
```
Which means the file will be closed as soon as the assignment is done (due to the `with` statement). This ends with an error at the `temp_file.write("text")` (due to the file being closed). 

My point is that the resulting code is quite useless as it only creates the file. Nothing else can be done with it since the file is closed almost as soon as it is created.

Reply:
The codemod isn't implemented to change your test example. It short circuits early because it doesn't match the pattern I've asked it for which is much more specific than what you have. In fact, there are similar tests all under the xfailed class, but I've added your test example as a "no_change" test to demonstrate this.

Reply:
Eventually I think we can/should make this codemod smarter and allow it to function much the same way as the resource leak codemod. But for now my understanding is that it will simply replace usages of `tempfile.mktemp()` in order to get a temporary filename.

Can we make sure to add an issue for the more sophisticated behavior if it doesn't already exist?

Reply:
Exactly! I communicated this at our previous meeting, this is iterative and this PR addresses only the first easiest cases as a way to demonstrate that we're aware the original codemod was just wrong. Will ticket the rest.